apis/apiextensions/v1/composition_types.go

```diff
@@ -134,6 +134,17 @@ func TypeReferenceTo(gvk schema.GroupVersionKind) TypeReference {
 // ComposedTemplate is used to provide information about how the composed resource
 // should be processed.
 type ComposedTemplate struct {
+	// TODO(negz): TemplateName should be a required field in v2 of this API.
+
+	// A TemplateName uniquely identifies this resource template within its
+	// Composition. Template names are strongly recommended. Either all or no
+	// templates within a Composition must be named. Compositions that use
+	// anonymous templates match templates to composed resources by order, and
+	// must therefore treat the order and length of their spec.resources array
+	// as immutable.
+	// +optional
+	TemplateName *string `json:"templateName,omitempty"`
```

Comment:
`baseName` seems like a good option. I guess technically the base is only part of thing this identifies (i.e. it's really naming the sum of the base, templates, etc) but I'm happy to avoid leaking the term 'template' to users.

I prefer to avoid simply `name` because I have a hunch that folks could easily assume it controls the `metadata.name` of the rendered resource, rather than just identifying the template/base.

Reply:
Urgh naming is hard. I never really loved `templateName` because I don't like admitting that this is a "template" (which always makes me think of moustache style text templates). `baseName` felt strange when I tried it because it names more than just the base. So I've given up and just gone with `name`, which is probably [most in line with API conventions](https://github.com/kubernetes/community/blob/7af513b2ab0625af7da8ce39b77945fc3fa8104c/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps) anyhow. Hopefully I'm wrong and API consumers won't find this too confusing.